src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2.scala

```diff
@@ -0,0 +1,217 @@
+    join.copy(
+      left = updateIfSupported(join.left, join.condition.get),
+      right = updateIfSupported(join.right, join.condition.get))
+  }
+
+  private def updateIfSupported(plan: LogicalPlan, condition: Expression): LogicalPlan = {
+    // Either left or right or both sides should contain half of references.
+    // All references for any side, either left or right, should all come from the same leaf node.
+    //
+    // E.g. if join condition refers to columns A, B, C, D, two columns should come from left, 2
+    // from right. This would be by default true. Nothing to do in this part. Let's say A,B
+    // come from left. C,D come from right. The requirement is both A and B should come from the
+    // same leaf node on left. Same for C and D. Both should come from same leaf node on right.
+    require(condition.references.forall(_.isInstanceOf[AttributeReference]))
+    val joinCols =
+      extractConditions(condition).flatMap(_.references.map(_.asInstanceOf[AttributeReference]))
+    val eligibleBaseRelations = plan.collectLeaves().filter {
+      case relation: LogicalRelation => relation.output.exists(contains(joinCols, _))
+      case _ => false
+    }
+
+    eligibleBaseRelations match {
+      case Seq(r: LogicalRelation)
+          if r.output.toSet.count(c => contains(joinCols, c)) * 2 == joinCols.size =>
+        updateIndex(plan, joinCols, plan.output)
+      case _ => plan
+    }
+  }
+
+  /**
+   * Update the plan by replacing source data with indexes. To make sure the index covers all
+   * required columns, traverse down the plan and collect all referenced columns from all nodes
+   * starting from top of the plan to the bottom.
+   *
+   * Please note: we can't use plan.transform(...) here because we have to collect referenced
+   * columns from every node we traverse till the end.
+   * @param plan Original plan.
+   * @param joinCols All referenced join columns from join condition.
+   * @param requiredCols Superset of all columns which are referenced in the plan. If an index
+   *                     is chosen to replace a base relation `r`, it should contain all the
+   *                     columns from this `requiredCols`, which directly belong to `r`.
+   * @return Logical plan with indexes if applicable indexes are found, else the original plan.
+   */
+  private def updateIndex(
+      plan: LogicalPlan,
+      joinCols: Seq[AttributeReference],
+      requiredCols: Seq[Attribute]): LogicalPlan = {
+    plan match {
+      case p: LogicalRelation => setIndexes(p, joinCols, requiredCols)
+      case p: LeafNode => p
+      case p: LogicalPlan =>
+        p.withNewChildren {
+          p.children.map { child =>
+            updateIndex(child, joinCols, p.references.toSeq ++ requiredCols)
+          }
+        }
+    }
+  }
+
+  private def eligible(l: LogicalPlan, r: LogicalPlan, condition: Expression): Boolean = {
+    !isBroadcastJoin(l, r) && isJoinConditionSupported(condition)
```

Comment:
Could you please add something similar to `isPlanModified` ?https://github.com/microsoft/hyperspace/blob/master/src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala#L104

And does v2 join rule support non-linear child plans?

Reply:
thanks @sezruby ,added at logical relation level. please have a look.